crypto/context.c

```diff
@@ -120,18 +113,46 @@ static int context_deinit(OPENSSL_CTX *ctx)
 }
 
 #ifndef FIPS_MODULE
-void openssl_ctx_default_deinit(void)
+/* The default default context */
+static OPENSSL_CTX default_context_int;
+
+static CRYPTO_ONCE default_context_init = CRYPTO_ONCE_STATIC_INIT;
+static CRYPTO_THREAD_LOCAL default_context_thread_local;
+
+DEFINE_RUN_ONCE_STATIC(default_context_do_init)
 {
-    context_deinit(default_context);
+    return CRYPTO_THREAD_init_local(&default_context_thread_local, NULL)
+        && context_init(&default_context_int);
 }
 
-static CRYPTO_ONCE default_context_init = CRYPTO_ONCE_STATIC_INIT;
-DEFINE_RUN_ONCE_STATIC(do_default_context_init)
+static OPENSSL_CTX *get_default_context(void)
 {
-    if (context_init(&default_context_int))
-        default_context = &default_context_int;
+    OPENSSL_CTX *current_defctx;
 
-    return 1;
+    if (!RUN_ONCE(&default_context_init, default_context_do_init))
+        return NULL;
+
+    current_defctx = CRYPTO_THREAD_get_local(&default_context_thread_local);
+    if (current_defctx == (OPENSSL_CTX *)-1)
```

Comment:
Why the -1 cast and test here.  Doesn't `CRYPTO_THREAD_get_local()` return NULL if not found?

Reply:
I copied code from elsewhere, and misremembered how that was done...  It infuriates me, though, that NULL wouldn't be a valid value.

Reply:
Fixed, I simply removed this check

Reply:
NULL is fine here.  I thought this through when I asked about the cast.